sw/device/tests/usbdev_pullup_test.c

```diff
@@ -0,0 +1,214 @@
+// desired.
+static status_t config_set(bool pinflip) {
+  dif_usbdev_config_t config = {
+      .have_differential_receiver = kDifToggleEnabled,
+      .use_tx_d_se0 = kDifToggleDisabled,
+      .single_bit_eop = kDifToggleDisabled,
+      .pin_flip = dif_bool_to_toggle(pinflip),
+      .clock_sync_signals = kDifToggleEnabled,
+  };
+
+  TRY(dif_usbdev_configure(&usbdev, &buffer_pool, config));
+
+  return OK_STATUS();
+}
+
+// Wait with timeout until the VBUS/SENSE signal is in the expected state.
+static status_t vbus_wait(bool expected, uint32_t timeout_micros) {
+  ibex_timeout_t timeout = ibex_timeout_init(timeout_micros);
+  do {
+    // Read the current state of VBUS.
+    bool vbus;
+    TRY(dif_usbdev_status_get_sense(&usbdev, &vbus));
+    if (vbus == expected) {
+      return OK_STATUS();
+    }
+  } while (!ibex_timeout_check(&timeout));
+
+  return UNAVAILABLE();
+}
+
+// Wait with timeout for the specified USB data line to be in the expected
+// state.
+static status_t line_wait(bool dp, bool expected, uint32_t timeout_micros) {
+  ibex_timeout_t timeout = ibex_timeout_init(timeout_micros);
+  do {
+    // Sense the current state of the pins.
+    dif_usbdev_phy_pins_sense_t status;
+    TRY(dif_usbdev_get_phy_pins_status(&usbdev, &status));
+
+    if ((dp && status.rx_dp == expected) || (!dp && status.rx_dn == expected)) {
+      return OK_STATUS();
+    }
+  } while (!ibex_timeout_check(&timeout));
+
+  return UNAVAILABLE();
+}
+
+// Delay for the specified number of microseconds, with user reporting for
+// appropriate targets.
+static status_t delay(bool prompt, uint32_t timeout_micros) {
+  if (prompt) {
+    LOG_INFO("Delaying...");
+  }
+  busy_spin_micros(timeout_micros);
+
+  return OK_STATUS();
+}
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  CHECK(kDeviceType == kDeviceSimDV || kDeviceType == kDeviceSimVerilator ||
+            kDeviceType == kDeviceFpgaCw310,
+        "This test is not expected to run on platforms other than the "
+        "Verilator/DV simulation or CW310 FPGA. It needs either the DPI model "
+        "or a VBUS/SENSE connection to a physical host");
+
+  uint32_t timeout_micros = 1000u;
+  uint32_t delay_micros = 1u;
+  bool can_flip = true;
+  bool prompt = false;
+
+  switch (kDeviceType) {
+    // In simulation the DPI model connects VBUS shortly after reset and
+    // prolonged delays when asserting or deasserting pull ups are wasteful.
+    case kDeviceSimDV:
+    case kDeviceSimVerilator:
+      break;
+    default:
+      // FPGA platforms where user intervention may be required.
+      timeout_micros = 30 * 1000 * 1000u;
+      // A short delay here permits the activity of the host controller to be
+      // observed (eg. dmesg -w on a Linux host).
+      delay_micros = 2 * 1000 * 1000u;
+      can_flip = false;
+      prompt = true;
+
+      LOG_INFO("Running USBDEV_VBUS test");
+      break;
+  }
+
+  // Ensure that the VBUS/SENSE signal is routed through to the usbdev
```

Comment:
Okay, thanks. I've changed the switch statement in this instance. I prefer to use switch rather than if/else when selecting from a set of values for a few reasons - it's more extensible/efficient when additional values are added later, by avoiding the tendency of code to grow over time into ever more complex if/else ladders involving repeated checks. At run time only one value can match, of course.
Generally, it also offers the 'default' clause for catching code that hasn't been updated to support a new target.

In this case we're perhaps unlikely to acquire new simulation platforms any time soon, since 'SimDV' is already an umbrella term for multiple simulators. We shall of course need to support the CW340 very soon, and there will surely be SiVal test platforms appearing soon too.

Re moving code into lib/testing I think that should happen when knowledge of the SiVal environment and requirements has crystallized, but for now I want to keep each SiVal test as minimal and self-contained as possible.